### PR TITLE
🎨 Optimize project listing queries: replace anti-join and GROUP BY with EXISTS/NOT EXISTS

### LINKED ISSUE #9027: [Maintenance]: Project listing is slow after merge of deployments
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Planned changes

After merging of the two deployments the sleeper tests in s4llite (i.e. this is taking place on https://sim4life.io/ within the s4llite product) fail because the listing of projects is too slow. I investigated this using tracing from which it was clear that the bottleneck was the PostgreSQL db and from the aws performance insights I dug out the following two queries which are very slow
```sql
EXPLAIN ANALYZE
SELECT
  count(*) AS count_1
FROM (
  SELECT
    p.id               AS id,
    p.type             AS type,
    p.template_type    AS template_type,
    p.uuid             AS uuid,
    p.name             AS name,
    p.description      AS description,
    p.thumbnail        AS thumbnail,
    p.prj_owner        AS prj_owner,
    p.creation_date    AS creation_date,
    p.last_change_date AS last_change_date,
    p.ui               AS ui,
    p.classifiers      AS classifiers,
    p.dev              AS dev,
    p.quality          AS quality,
    p.published        AS published,
    p.hidden           AS hidden,
    p.workspace_id     AS workspace_id,
    p.trashed          AS trashed,
    p.trashed_by       AS trashed_by,
    p.trashed_explicitly AS trashed_explicitly,
    p.product_name     AS product_name,
    p.workbench        AS workbench,
    pf.folder_id       AS folder_id
  FROM projects AS p
  JOIN (
    SELECT
      pg.project_uuid AS project_uuid
    FROM project_to_groups AS pg
    WHERE
      pg.read
      AND pg.gid IN (63, 1, 4, 22, 11, 30, 2135745, 2135744)
    GROUP BY
      pg.project_uuid
  ) AS my_access_rights_subquery
    ON p.uuid = my_access_rights_subquery.project_uuid
  LEFT OUTER JOIN projects_to_folders AS pf
    ON pf.project_uuid = p.uuid
   AND pf.user_id = 52
  WHERE
    p.workspace_id IS NULL
    AND p.product_name = 's4llite'
    AND p.type = 'STANDARD'
    AND p.hidden IS false
    AND p.trashed IS NULL
    AND pf.folder_id IS NULL
) AS anon_1;
```

```sql
EXPLAIN ANALYZE
SELECT
  projects.id,
  projects.type,
  projects.template_type,
  projects.uuid,
  projects.name,
  projects.description,
  projects.thumbnail,
  projects.prj_owner,
  projects.creation_date,
  projects.last_change_date,
  projects.ui,
  projects.classifiers,
  projects.dev,
  projects.quality,
  projects.published,
  projects.hidden,
  projects.workspace_id,
  projects.trashed,
  projects.trashed_by,
  projects.trashed_explicitly,
  projects.product_name,
  projects.workbench,
  projects_to_folders.folder_id
FROM projects
JOIN (
  SELECT project_to_groups.project_uuid AS project_uuid
  FROM project_to_groups
  WHERE project_to_groups.read
    AND project_to_groups.gid IN (63, 1, 4, 22, 11, 30, 2135745, 2135744)
  GROUP BY project_to_groups.project_uuid
) AS my_access_rights_subquery
  ON projects.uuid = my_access_rights_subquery.project_uuid
LEFT OUTER JOIN projects_to_folders
  ON projects_to_folders.project_uuid = projects.uuid
 AND projects_to_folders.user_id = 52
WHERE projects.workspace_id IS NULL
  AND projects.product_name = 's4llite'
  AND projects.type = 'STANDARD'
  AND projects.hidden IS false
  AND projects.trashed IS NULL
  AND projects_to_folders.folder_id IS NULL
ORDER BY projects.last_change_date DESC, projects.id
LIMIT 100 OFFSET 0;
```

After investigating together with Perplexity and @matusdrobuliak66 it is clear that the main bottleneck is in the `WHERE` clause

```sql
WHERE projects.workspace_id IS NULL
  AND projects.product_name = 's4llite'
  AND projects.type = 'STANDARD'
  AND projects.hidden IS false
  AND projects.trashed IS NULL
  AND projects_to_folders.folder_id IS NULL

```
This clause concerns data from the `projects` table as well as data from the `projects_to_folder` table. For that reason the filtering is only performed **after** the join. When removing "  AND projects_to_folders.folder_id IS NULL" from this clause, the execution time drops from 35seconds to 88miliseconds (this was tested in a test db on aws created before the merge of the deployments - in order to not mess to much with prod). Hence, probably the way `projects_to_folders` enters the query should be changed.

### PR DESCRIPTION
**Pull request by matusdrobuliak66**
## What do these changes do?

Rewrites the project listing SQL queries to eliminate two performance bottlenecks:

1. Access rights: Replaced GROUP BY subquery joined via INNER JOIN with a correlated EXISTS check in WHERE. The planner can now stop at the first matching row instead of materialising a grouped intermediate result.

2. Folder filtering: Split the single LEFT JOIN ... WHERE folder_id IS NULL pattern into three scope-specific strategies:
    - ROOT: NOT EXISTS subquery (no join) — avoids the costly anti-join pattern that caused slow root listings **<-- This is what we were hitting here** https://github.com/ITISFoundation/osparc-simcore/issues/9027
    - SPECIFIC: INNER JOIN on exact [(user_id, folder_id)](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) — simple equi-join
    - ALL: LEFT OUTER JOIN (no filter) — unchanged, used by full-depth search

## Related issue/s

- fixes https://github.com/ITISFoundation/osparc-simcore/issues/9027

## How to test

All existing tests pass (55 tests across listing, folders, and workspaces suites). Added 4 new edge-case tests:
- Multiple folders isolation (projects don't leak between folders)
- Pagination correctness after folder split
- Search respects folder scope
- Round-trip: project returns to root after folder assignment removal

### CODE REVIEW

**Review by giancarloromeo — APPROVED** [review]
👍

File: services/web/server/tests/unit/with_dbs/02/test_projects_crud_handlers__list_with_query_params.py [hunk 1]
```diff
@@ -671,3 +671,355 @@ async def test_list_and_patch_projects_with_template_type(  # noqa: PLR0915
...
+    with postgres_db.connect() as con:
+        con.execute(projects_to_folders.delete())
+        con.execute(folders_v2.delete())
+
+
+@pytest.fixture()
+def _create_folder(
+    postgres_db: sa.engine.Engine,
+    logged_user: UserInfoDict,
+) -> Iterator[Callable[[str], FolderID]]:
```
Comment (GitHK) [inline services/web/server/tests/unit/with_dbs/02/test_projects_crud_handlers__list_with_query_params.py]: I have a feeling this already exists, please try to find a similar fixture and reuse/refactor it

**Review by bisgaard-itis — APPROVED** [review]
Very interesting, thanks a lot. I really like your plan.md.
Did you generate the SQL query you would not get for the playwright user and try it out in the production db? If not I suggest you do so, to check you get the expected performance improvements.

**Review by GitHK — APPROVED** [review]
Thanks!

**Review by pcrespov — APPROVED** [review]
